curvefs/src/tools/curvefs_tool_factory.cpp

```diff
@@ -0,0 +1,54 @@
+/*
+ *  Copyright (c) 2021 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Project: curve
+ * Created Date: 2021-09-14
+ * Author: chengyi01
+ */
+
+#include "curvefs/src/tools/curvefs_tool_factory.h"
+
+namespace curvefs {
+namespace tools {
+
+CurvefsToolFactory::CurvefsToolFactory() {
+    // version
+    RegisterCurvefsTool(std::string(kVersionCmd),
+                        CurvefsToolCreator<version::VersionTool>::Create);
+
+    // umountfs
+    RegisterCurvefsTool(std::string(kUmountCmd),
+                        CurvefsToolCreator<umountfs::UmountfsTool>::Create);
+}
+
+std::shared_ptr<CurvefsTool> CurvefsToolFactory::GenerateCurvefsTool(
+    const std::string& command) {
+    auto search = command2creator_.find(command);
+    if (search != command2creator_.end()) {
+        return search->second();
+    }
+    return nullptr;
+}
+
+void CurvefsToolFactory::RegisterCurvefsTool(
+    const std::string& command,
+    const std::function<std::shared_ptr<CurvefsTool>()>& function) {
+    command2creator_.insert({command, function});
```

Comment:
check if the command already exists?

Reply:
> check if the command already exists?

I thought about this problem, but this kind of check is best placed at compile time, such as using conditional compilation, but this scheme still cannot prevent the abuse of the register function. As for making the register function private, it will reduce flexibility, such as generating different tools according to conditions, but in this case, the initialization of the command should be placed in Init().